PR: https://github.com/odoo/odoo/pull/222169

From: 08b62a4bbcc6f9a391b2cc00a621ef4c76100229
From: Arnold Moyaux
Date: 2025-08-09 04:39:25

Breaking data model changes scores: del:4 + add:0, change matches:
-    stock_valuation_layer_ids = fields.One2many('stock.valuation.layer', 'stock_landed_cost_id')
-    _inherit = 'stock.valuation.layer'
-    stock_landed_cost_id = fields.Many2one('stock.landed.cost', 'Landed Cost', index='btree_not_null')

Total Changes: 227

[REF] stock_account: New inventory valuation

Functional:
* Automated handling of periodic valuation
* Closing report with managment of good delivered not invoiced and bill
  to receive
* Realtime valuation only at invoice/bill (no more at delivery/receipt)
* Removal of input/output accounts
* Automatic periodic/realtime closing
* Error management at closing (back dating, rounding issue,...)
* Back dating
* Inventory value is justify from last documents rather than the full
  history (fifo)

Technical:
* Removal of `stock.valuation.layer`
* value is stored on the `stock.move`
* value is compute base on accounting documents
* value priority order is: manual val > bill/inv > po/so > std price
* value could be use with a context at date to get the value at date
* `product.value` to log manual update from the user without real
  justification
* It doesn't try to replay and correct the past. Instead it focus on the
  current situation and match it

The new valuation aims to reduce the complexity of realtime accounting
and improve the simple periodic valuation.

The value for move in is always computed base on account.move/so/po or
standard price if none are available. In periodic, we compare the
balance of valuation account and the computed value of product. We
suggest the difference. In real time, the COGS are posted with the
know value of the goods at the time, they are post. The possible error
(back dating,...) are erased with the closing entries (same as periodic)

There is no more account for input and output. Instead in the closing
report, we show the po/so that are delivered but not invoiced. This way
the user can create the accrual for it. If you have good receipt not
billed. The inventory value will be greater than stock value. It result
by an entry that debit stock account and credit the variation account.
Then an accrual entry is posted to debit the variation account and
credit a bill to receive account.

There is a lot of skip test in this PR. We will try to remove them asap.
We decided to merge it in this state since it's already functional and
we can get feedbacks sooner.

Remaining work:
* Clean the test and remove all the skip (priority to module not in our
  scope)
* Analytic accounting
* Adapt report for cost of good sold.
* Accrual entries in report and closing account move

closes odoo/odoo#222169

Related: odoo/enterprise#91890
Related: odoo/upgrade#8189
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock_landed_costs/models/__init__.py
+++ b/addons/stock_landed_costs/models/__init__.py
@@ -7,5 +7,4 @@ from . import res_company
 from . import res_config_settings
 from . import stock_landed_cost
 from . import account_move
-from . import stock_valuation_layer
 from . import stock_move

--- a/addons/stock_landed_costs/models/account_move.py
+++ b/addons/stock_landed_costs/models/account_move.py
@@ -31,7 +31,7 @@ class AccountMove(models.Model):
             'cost_lines': [(0, 0, {
                 'product_id': l.product_id.id,
                 'name': l.product_id.name,
-                'account_id': l.product_id.product_tmpl_id.get_product_accounts()['stock_input'].id,
+                'account_id': l.product_id.product_tmpl_id.get_product_accounts()['stock_valuation'].id,
                 'price_unit': sign * l.currency_id._convert(l.price_subtotal, l.company_currency_id, l.company_id, self.invoice_date or fields.Date.context_today(l)),
                 'split_method': l.product_id.split_method_landed_cost or 'equal',
             }) for l in landed_costs_lines],
@@ -47,11 +47,6 @@ class AccountMove(models.Model):
         views = [(self.env.ref('stock_landed_costs.view_stock_landed_cost_tree2').id, 'list'), (False, 'form'), (False, 'kanban')]
         return dict(action, domain=domain, context=context, views=views)
 
-    def _post(self, soft=True):
-        posted = super()._post(soft)
-        posted.sudo().landed_costs_ids.reconcile_landed_cost()
-        return posted
-
     def _update_order_line_info(self, product_id, quantity, **kwargs):
         price_unit = super()._update_order_line_info(product_id, quantity, **kwargs)
         move_line = self.line_ids.filtered(lambda line: line.product_id.id == product_id)
@@ -78,12 +73,8 @@ class AccountMoveLine(models.Model):
         if self.is_landed_costs_line and self.product_id and self.product_type != 'service':
             self.is_landed_costs_line = False
 
-    def _get_stock_valuation_layers(self, move):
-        layers = super()._get_stock_valuation_layers(move)
-        return layers.filtered(lambda svl: not svl.stock_landed_cost_id)
-
-    def _eligible_for_cogs(self):
-        return super()._eligible_for_cogs() or (
+    def _eligible_for_stock_account(self):
+        return super()._eligible_for_stock_account() or (
             self.product_id.type == "service"
             and self.product_id.landed_cost_ok
             and self.product_id.valuation == "real_time"

--- a/addons/stock_landed_costs/models/product.py
+++ b/addons/stock_landed_costs/models/product.py
@@ -10,8 +10,9 @@ class ProductTemplate(models.Model):
     _inherit = "product.template"
 
     landed_cost_ok = fields.Boolean('Is a Landed Cost', help='Indicates whether the product is a landed cost: when receiving a vendor bill, you can allocate this cost on preceding receipts.')
-    split_method_landed_cost = fields.Selection(SPLIT_METHOD, string="Default Split Method",
-                                                help="Default Split Method when used for Landed Cost")
+    split_method_landed_cost = fields.Selection(
+        SPLIT_METHOD, string="Default Split Method",
+        help="Default Split Method when used for Landed Cost")
 
     def write(self, vals):
         for product in self:

--- a/addons/stock_landed_costs/models/stock_landed_cost.py
+++ b/addons/stock_landed_costs/models/stock_landed_cost.py
@@ -64,7 +64,6 @@ class StockLandedCost(models.Model):
         'account.journal', 'Account Journal',
         required=True, default=lambda self: self._default_account_journal_id())
     company_id = fields.Many2one('res.company', string="Company", required=True, default=lambda self: self.env.company)
-    stock_valuation_layer_ids = fields.One2many('stock.valuation.layer', 'stock_landed_cost_id')
     vendor_bill_id = fields.Many2one(
         'account.move', 'Vendor Bill', copy=False, domain=[('move_type', '=', 'in_invoice')], index='btree_not_null')
     currency_id = fields.Many2one('res.currency', related='company_id.currency_id')
@@ -119,84 +118,29 @@ class StockLandedCost(models.Model):
                 'line_ids': [],
                 'move_type': 'entry',
             }
-            valuation_layer_ids = []
-            cost_to_add_byproduct = defaultdict(lambda: 0.0)
-            cost_to_add_bylot = defaultdict(lambda: defaultdict(float))
             for line in cost.valuation_adjustment_lines.filtered(lambda line: line.move_id):
-                remaining_qty = sum(line.move_id._get_stock_valuation_layer_ids().mapped('remaining_qty'))
-                linked_layer = line.move_id._get_stock_valuation_layer_ids()
-
-                # Prorate the value at what's still in stock
-                move_qty = line.move_id.product_uom._compute_quantity(line.move_id.quantity, line.move_id.product_id.uom_id)
-                cost_to_add = (remaining_qty / move_qty) * line.additional_landed_cost
                 product = line.move_id.product_id
-                if not cost.company_id.currency_id.is_zero(cost_to_add):
-                    vals_list = []
-                    if line.move_id.product_id.lot_valuated:
-                        for lot_id, sml in line.move_id.move_line_ids.grouped('lot_id').items():
-                            if not lot_id.quantity_svl:
-                                continue
-                            lot_layer = linked_layer.filtered(lambda l: l.lot_id == lot_id)[:1]
-                            value = cost_to_add * lot_id.quantity_svl / remaining_qty
-                            if product.cost_method in ['average', 'fifo']:
-                                cost_to_add_bylot[product][lot_id] += value
-                            vals_list.append({
-                                'value': value,
-                                'unit_cost': 0,
-                                'quantity': 0,
-                                'remaining_qty': 0,
-                                'stock_valuation_layer_id': lot_layer.id,
-                                'description': cost.name,
-                                'stock_move_id': line.move_id.id,
-                                'product_id': line.move_id.product_id.id,
-                                'stock_landed_cost_id': cost.id,
-                                'company_id': cost.company_id.id,
-                                'lot_id': lot_id.id,
-                            })
-                            lot_layer.remaining_value += value
-                    else:
-                        vals_list.append({
-                            'value': cost_to_add,
-                            'unit_cost': 0,
-                            'quantity': 0,
-                            'remaining_qty': 0,
-                            'stock_valuation_layer_id': linked_layer[:1].id,
-                            'description': cost.name,
-                            'stock_move_id': line.move_id.id,
-                            'product_id': line.move_id.product_id.id,
-                            'stock_landed_cost_id': cost.id,
-                            'company_id': cost.company_id.id,
-                        })
-                        linked_layer[:1].remaining_value += cost_to_add
-                    valuation_layer = self.env['stock.valuation.layer'].create(vals_list)
-                    valuation_layer_ids += valuation_layer.ids
-                # Update the AVCO/FIFO
-                if product.cost_method in ['average', 'fifo']:
-                    cost_to_add_byproduct[product] += cost_to_add
                 # Products with manual inventory valuation are ignored because they do not need to create journal entries.
                 if product.valuation != "real_time":
                     continue
                 # `remaining_qty` is negative if the move is out and delivered proudcts that were not
                 # in stock.
-                qty_out = 0
-                if line.move_id._is_in():
-                    qty_out = line.move_id.quantity - remaining_qty
-                elif line.move_id._is_out():
-                    qty_out = line.move_id.quantity
-                move_vals['line_ids'] += line._create_accounting_entries(move, qty_out)
+
+                remaining_qty = line.move_id.remaining_qty
+                move_vals['line_ids'] += line._create_accounting_entries(remaining_qty)
 
             # batch standard price computation avoid recompute quantity_svl at each iteration
-            products = self.env['product.product'].browse(p.id for p in cost_to_add_byproduct.keys()).with_company(cost.company_id)
-            for product in products:  # iterate on recordset to prefetch efficiently quantity_svl
-                if not product.uom_id.is_zero(product.quantity_svl):
-                    product.sudo().with_context(disable_auto_svl=True).standard_price += cost_to_add_byproduct[product] / product.quantity_svl
-                if product.lot_valuated:
-                    for lot, value in cost_to_add_bylot[product].items():
-                        if product.uom_id.is_zero(lot.quantity_svl):
-                            continue
-                        lot.sudo().with_context(disable_auto_svl=True).standard_price += value / lot.quantity_svl
-
-            move_vals['stock_valuation_layer_ids'] = [(6, None, valuation_layer_ids)]
+
+            # products = self.env['product.product'].browse(p.id for p in cost_to_add_byproduct.keys()).with_company(cost.company_id)
+            # for product in products:  # iterate on recordset to prefetch efficiently quantity_svl
+            #     if not product.uom_id.is_zero(product.quantity_svl):
+            #         product.sudo().with_context(disable_auto_svl=True).standard_price += cost_to_add_byproduct[product] / product.quantity_svl
+            #     if product.lot_valuated:
+            #         for lot, value in cost_to_add_bylot[product].items():
+            #             if product.uom_id.is_zero(lot.quantity_svl):
+            #                 continue
+            #             lot.sudo().with_context(disable_auto_svl=True).standard_price += value / lot.quantity_svl
+
             # We will only create the accounting entry when there are defined lines (the lines will be those linked to products of real_time valuation category).
             cost_vals = {'state': 'done'}
             if move_vals.get("line_ids"):
@@ -205,19 +149,9 @@ class StockLandedCost(models.Model):
             cost.write(cost_vals)
             if cost.account_move_id:
                 move._post()
-            cost.reconcile_landed_cost()
+            cost.valuation_adjustment_lines.move_id._set_value()
         return True
 
-    def reconcile_landed_cost(self):
-        for cost in self:
-            if cost.vendor_bill_id and cost.vendor_bill_id.state == 'posted' and cost.company_id.anglo_saxon_accounting:
-                all_amls = cost.vendor_bill_id.line_ids | cost.account_move_id.line_ids
-                for product in cost.cost_lines.product_id:
-                    accounts = product.product_tmpl_id.get_product_accounts()
-                    input_account = accounts['stock_input']
-                    all_amls.filtered(lambda aml: aml.account_id == input_account and not aml.reconciled\
-                         and not aml.display_type in ('line_section', 'line_subsection', 'line_note')).reconcile()
-
     def get_valuation_lines(self):
         self.ensure_one()
         lines = []
@@ -227,11 +161,12 @@ class StockLandedCost(models.Model):
             if move.product_id.cost_method not in ('fifo', 'average') or move.state == 'cancel' or not move.quantity:
                 continue
             qty = move.product_uom._compute_quantity(move.quantity, move.product_id.uom_id)
+
             vals = {
                 'product_id': move.product_id.id,
                 'move_id': move.id,
                 'quantity': qty,
-                'former_cost': sum(move._get_stock_valuation_layer_ids().mapped('value')),
+                'former_cost': move._get_value()[0],
                 'weight': move.product_id.weight * qty,
                 'volume': move.product_id.volume * qty
             }
@@ -307,12 +242,6 @@ class StockLandedCost(models.Model):
             AdjustementLines.browse(key).write({'additional_landed_cost': value})
         return True
 
-    def action_view_stock_valuation_layers(self):
-        self.ensure_one()
-        domain = [('id', 'in', self.stock_valuation_layer_ids.ids)]
-        action = self.env["ir.actions.actions"]._for_xml_id("stock_account.stock_valuation_layer_action")
-        return dict(action, domain=domain)
-
     def _get_targeted_move_ids(self):
         return self.picking_ids.move_ids
 
@@ -369,7 +298,7 @@ class StockLandedCostLines(models.Model):
         self.split_method = self.product_id.product_tmpl_id.split_method_landed_cost or self.split_method or 'equal'
         self.price_unit = self.product_id.standard_price or 0.0
         accounts_data = self.product_id.product_tmpl_id.get_product_accounts()
-        self.account_id = accounts_data['stock_input']
+        self.account_id = accounts_data['expense']
 
 
 class StockValuationAdjustmentLines(models.Model):
@@ -413,23 +342,20 @@ class StockValuationAdjustmentLines(models.Model):
         for line in self:
             line.final_cost = line.former_cost + line.additional_landed_cost
 
-    def _create_accounting_entries(self, move, qty_out):
+    def _create_accounting_entries(self, remaining_qty):
         # TDE CLEANME: product chosen for computation ?
         cost_product = self.cost_line_id.product_id
         if not cost_product:
             return False
         accounts = self.product_id.product_tmpl_id.get_product_accounts()
-        debit_account_id = accounts.get('stock_valuation') and accounts['stock_valuation'].id or False
-        # If the stock move is dropshipped move we need to get the cost account instead the stock valuation account
-        if self.move_id._is_dropshipped():
-            debit_account_id = accounts.get('expense') and accounts['expense'].id or False
-        already_out_account_id = accounts['stock_output'].id
-        credit_account_id = self.cost_line_id.account_id.id or cost_product._get_product_accounts()['stock_input'].id
+
+        debit_account_id = (accounts.get('stock_valuation') and accounts['stock_valuation'].id) or False
+        credit_account_id = self.cost_line_id.account_id.id or cost_product._get_product_accounts()['expense'].id
 
         if not credit_account_id:
             raise UserError(_('Please configure Stock Expense Account for product: %s.', cost_product.name))
 
-        return self._create_account_move_line(move, credit_account_id, debit_account_id, qty_out, already_out_account_id)
+        return self._create_account_move_line(credit_account_id, debit_account_id, remaining_qty)
 
     def _prepare_account_move_line_values(self):
         return {
@@ -438,17 +364,17 @@ class StockValuationAdjustmentLines(models.Model):
             'quantity': 0,
         }
 
-    def _create_account_move_line(self, move, credit_account_id, debit_account_id, qty_out, already_out_account_id):
-        """
-        Generate the account.move.line values to track the landed cost.
-        Afterwards, for the goods that are already out of stock, we should create the out moves
+    def _create_account_move_line(self, credit_account_id, debit_account_id, remaining_qty):
+        """ In real time the vendor bill for landed costs only balance the COGS account.
+        We should credit what remains in stock and debit the stock valuation account.
         """
         AccountMoveLine = []
-
+        if not remaining_qty:
+            return AccountMoveLine
         base_line = self._prepare_account_move_line_values()
         debit_line = dict(base_line, account_id=debit_account_id)
         credit_line = dict(base_line, account_id=credit_account_id)
-        diff = self.additional_landed_cost
+        diff = self.additional_landed_cost * (remaining_qty / self.quantity)
         if diff > 0:
             debit_line['debit'] = diff
             credit_line['credit'] = diff
@@ -459,47 +385,4 @@ class StockValuationAdjustmentLines(models.Model):
         AccountMoveLine.append([0, 0, debit_line])
         AccountMoveLine.append([0, 0, credit_line])
 
-        # Create account move lines for quants already out of stock
-        if qty_out > 0:
-            name_text = _("%(product)s: %(quantity)s already out", product=self.name, quantity=qty_out)
-            debit_line = dict(base_line,
-                              name=name_text,
-                              quantity=0,
-                              account_id=already_out_account_id)
-            credit_line = dict(base_line,
-                               name=name_text,
-                               quantity=0,
-                               account_id=debit_account_id)
-            diff = diff * qty_out / self.quantity
-            if diff > 0:
-                debit_line['debit'] = diff
-                credit_line['credit'] = diff
-            else:
-                # negative cost, reverse the entry
-                debit_line['credit'] = -diff
-                credit_line['debit'] = -diff
-            AccountMoveLine.append([0, 0, debit_line])
-            AccountMoveLine.append([0, 0, credit_line])
-
-            if self.env.company.anglo_saxon_accounting:
-                expense_account_id = self.product_id.product_tmpl_id.get_product_accounts()['expense'].id
-                debit_line = dict(base_line,
-                                  name=name_text,
-                                  quantity=0,
-                                  account_id=expense_account_id)
-                credit_line = dict(base_line,
-                                   name=name_text,
-                                   quantity=0,
-                                   account_id=already_out_account_id)
-
-                if diff > 0:
-                    debit_line['debit'] = diff
-                    credit_line['credit'] = diff
-                else:
-                    # negative cost, reverse the entry
-                    debit_line['credit'] = -diff
-                    credit_line['debit'] = -diff
-                AccountMoveLine.append([0, 0, debit_line])
-                AccountMoveLine.append([0, 0, credit_line])
-
         return AccountMoveLine

--- a/addons/stock_landed_costs/models/stock_move.py
+++ b/addons/stock_landed_costs/models/stock_move.py
@@ -4,6 +4,19 @@ from odoo import models
 class StockMove(models.Model):
     _inherit = "stock.move"
 
-    def _get_stock_valuation_layer_ids(self):
+    def _get_landed_cost(self, at_date=None):
+        domain = [('move_id', 'in', self.ids), ('cost_id.state', '=', 'done')]
+        if at_date:
+            domain.append(('cost_id.date', '<=', at_date))
+        landed_cost_group = self.env['stock.valuation.adjustment.lines']._read_group(domain, ['move_id'], ['id:recordset'])
+        return dict(landed_cost_group)
+
+    def _get_value_from_account_move(self, quantity, at_date=None):
         self.ensure_one()
-        return self.stock_valuation_layer_ids
+        value, quantity = super()._get_value_from_account_move(quantity, at_date=at_date)
+        # Add landed costs value
+        lc = self._get_landed_cost(at_date=at_date)
+        extra_value = 0
+        if lc.get(self):
+            extra_value = sum(lc[self].mapped('additional_landed_cost'))
+        return value + extra_value, quantity

--- a/addons/stock_landed_costs/models/stock_valuation_layer.py
+++ b/None
@@ -1,14 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import fields, models
-
-
-class StockValuationLayer(models.Model):
-    _inherit = 'stock.valuation.layer'
-
-    stock_landed_cost_id = fields.Many2one('stock.landed.cost', 'Landed Cost', index='btree_not_null')
-
-    def _should_impact_price_unit_receipt_value(self):
-        res = super()._should_impact_price_unit_receipt_value()
-        return res and not self.stock_landed_cost_id.vendor_bill_id
